### PR TITLE
Add `regexp_like, improve docs and examples for `regexp_match`

### LINKED ISSUE #9102: Add regexp_like scalar function
### Is your feature request related to a problem or challenge?

Currently there is regexp_match and regexp_replace however there isn't a corresponding regexp_like function that could be used in when(..) dataframe method or sql case statements.

### Describe the solution you'd like

An implementation of regexp_like that matches the syntax and style of the [postgresql implementation](https://www.postgresql.org/docs/16/functions-string.html) as close as possible. Note that the Spark version of the [regexp_like function](https://spark.apache.org/docs/latest/api/sql/#regexp_like) is very similar but does not include any flags in the function signature.

Not all of the flags that [postgresql supports](https://www.postgresql.org/docs/16/functions-matching.html#POSIX-EMBEDDED-OPTIONS-TABLE) may be included in the initial implementation - likely just 'i' may be implemented.

It is noted that the implementation for the existing regexp_match currently resides in datafusion however with https://github.com/apache/arrow-rs/pull/5235 this functionality was moved into arrow_rs (but has yet to be removed from datafusion). The implementation for regexp_like may take a similar path - implement in datafusion first then move to arrow_rs if the community thinks that would be a good idea.

### Describe alternatives you've considered

You can use regexp_match and test the return list for empty to imitate this function however that is not the most performant way to implement this as that methodology cannot return immediately after the first match.

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by Omega359**
## Which issue does this PR close?

Closes #9102

## Rationale for this change

Add a new regexp_like function.

## What changes are included in this PR?

Code, tests, documentation, perf test

## Are these changes tested?

Yes,

## Are there any user-facing changes?

A new regexp_like function is available for use.

### CODE REVIEW

**Comment by Omega359** [comment]
FYI - the prettier error is not something I see on my machine
```
arrow-datafusion on  feature/regexp_like [$!] is 📦 v35.0.0 via 🦀 v1.75.0 on ☁️  (us-east-1)
❯ prettier --version
3.2.4

arrow-datafusion on  feature/regexp_like [$!] is 📦 v35.0.0 via 🦀 v1.75.0 on ☁️  (us-east-1)
❯ prettier -w docs/**/scalar_functions.md
docs/source/user-guide/sql/scalar_functions.md 356ms (unchanged)
```

File: datafusion-examples/examples/regexp.rs [hunk 1]
```diff
@@ -0,0 +1,300 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
```
Comment (comphead) [inline datafusion-examples/examples/regexp.rs:9]: IMHO I like adding examples for introduced functions, but we probably can do that on SQL only? like https://spark.apache.org/docs/latest/api/sql/index.html

File: datafusion/physical-expr/benches/regexp.rs [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+use rand::distributions::Alphanumeric;
+use rand::rngs::ThreadRng;
+use rand::seq::SliceRandom;
+use rand::Rng;
+
+use datafusion_physical_expr::regex_expressions::{
+    regexp_like, regexp_match, regexp_replace,
+};
+
+fn data(rng: &mut ThreadRng) -> StringArray {
```
Comment (comphead) [inline datafusion/physical-expr/benches/regexp.rs:35]: that bench is really needed taking into account how much probs we had with regex performance

File: datafusion/physical-expr/src/functions.rs [hunk 3]
```diff
@@ -608,6 +608,27 @@ pub fn create_physical_fun(
...
+            DataType::LargeUtf8 => {
+                let func = invoke_on_array_if_regex_expressions_feature_flag!(
+                    regexp_like,
+                    i64,
+                    "regexp_like"
+                );
+                make_scalar_function_inner(func)(args)
+            }
+            other => {
+                internal_err!("Unsupported data type {other:?} for function regexp_like")
```
Comment (comphead) [inline datafusion/physical-expr/src/functions.rs:629]:
```suggestion
                exec_err!("Unsupported data type {other:?} for function regexp_like")
```

File: docs/source/user-guide/sql/scalar_functions.md [hunk 4]
```diff
@@ -1785,10 +1893,10 @@ from_unixtime(expression)
 - [array_sort](#array_sort)
 - [array_cat](#array_cat)
 - [array_concat](#array_concat)
-- [array_contains](#array_contains)
```
Comment (comphead) [inline docs/source/user-guide/sql/scalar_functions.md:1788]: this is not related to PR?

File: datafusion/physical-expr/src/functions.rs (hunk 3 above)
Comment (Omega359) [reply datafusion/physical-expr/src/functions.rs:629]: There is quite a few instances in the code that have this pattern (I was mirroring the existing regexp_... functions there) - I was going to create a followup PR that fixes them across the whole code base

File: docs/source/user-guide/sql/scalar_functions.md (hunk 4 above)
Comment (Omega359) [reply docs/source/user-guide/sql/scalar_functions.md:1788]: Indeed you are correct, there are likely a couple of places where I fixed things as I saw them. Would you like those pulled into a separate PR?

File: datafusion-examples/examples/regexp.rs (hunk 1 above)
Comment (Omega359) [reply datafusion-examples/examples/regexp.rs:9]: When I started with DataFusion it was the examples that gave me the initial base to work from - especially since this projects was essentially my first time with Rust code. I personally think having both dataframe and SQL examples for the same thing to be quite helpful for those getting started who are not already 

I was actually thinking of spending some time (when I had it) to enhance the scalar docs to add examples where missing and to add whole new sections to the examples as above.

Comment (comphead) [reply datafusion-examples/examples/regexp.rs:9]: I'm thinking on doc tests? like in `datafusion/core/src/dataframe/mod.rs`

Comment (Omega359) [reply datafusion-examples/examples/regexp.rs:9]: That could work though it could be rather large doc. I'll update this PR in the next day or two with that change. I should likely do that with the to_timestamp example as well (in a separate PR of course).

Comment (Omega359) [reply datafusion-examples/examples/regexp.rs:9]: Done.

File: datafusion/physical-expr/src/functions.rs (hunk 3 above)
Comment (alamb) [reply datafusion/physical-expr/src/functions.rs:629]: That would be most appreciated. Cleaning up internal errors is very good improvement

File: datafusion-examples/README.md [hunk 5]
```diff
@@ -52,21 +52,22 @@ cargo run --example csv_sql
 - [`dataframe_output.rs`](examples/dataframe_output.rs): Examples of methods which write data out from a DataFrame
 - [`dataframe_in_memory.rs`](examples/dataframe_in_memory.rs): Run a query using a DataFrame against data in memory
 - [`deserialize_to_struct.rs`](examples/deserialize_to_struct.rs): Convert query results into rust structs using serde
-- [`expr_api.rs`](examples/expr_api.rs): Create, execute, simplify and anaylze `Expr`s
+- [`expr_api.rs`](examples/expr_api.rs): Create, execute, simplify and analyze `Expr`s
 - [`flight_sql_server.rs`](examples/flight/flight_sql_server.rs): Run DataFusion as a standalone process and execute SQL queries from JDBC clients
+- [`make_date.rs`](examples/make_date.rs): Examples of using the make_date function
```
Comment (alamb) [inline datafusion-examples/README.md:57]: Thank you 🙏

File: datafusion/proto/proto/datafusion.proto [hunk 6]
```diff
@@ -677,6 +677,7 @@ enum ScalarFunction {
   InStr = 132;
   MakeDate = 133;
   ArrayReverse = 134;
+  RegexpLike = 135;
```
Comment (alamb) [inline datafusion/proto/proto/datafusion.proto:680]: Doesn't this need RegexpMatch as well? I may be missing something

File: datafusion/sqllogictest/test_files/regexp.slt [hunk 7]
```diff
@@ -0,0 +1,303 @@
+# Licensed to the Apache Software Foundation (ASF) under one
```
Comment (alamb) [inline datafusion/sqllogictest/test_files/regexp.slt:1]: Thank you -- this is much better

File: datafusion/physical-expr/src/regex_expressions.rs [hunk 8]
```diff
@@ -54,7 +54,120 @@ macro_rules! fetch_string_arg {
...
+        3 => {
+            let values = as_generic_string_array::<T>(&args[0])?;
+            let regex = as_generic_string_array::<T>(&args[1])?;
+            let flags = as_generic_string_array::<T>(&args[2])?;
+
+            if flags.iter().any(|s| s == Some("g")) {
+                return plan_err!("regexp_like() does not support the \"global\" option");
+            }
+
+            let array = arrow_string::regexp::regexp_is_match_utf8(values, regex, Some(flags))
```
Comment (alamb) [inline datafusion/physical-expr/src/regex_expressions.rs:118]: 👍

File: datafusion/physical-expr/src/regex_expressions.rs [hunk 9]
```diff
@@ -54,7 +54,120 @@ macro_rules! fetch_string_arg {
...
+///     "b",
+///     regexp_like(vec![col("values"), col("patterns"), col("flags")])
+/// )?;
+/// // literals can be used as well with dataframe calls
+/// let df = df.with_column(
+///     "c",
+///     regexp_like(vec![lit("foobarbequebaz"), lit("(bar)(beque)")])
+/// )?;
+///
+/// df.show().await?;
```
Comment (alamb) [inline datafusion/physical-expr/src/regex_expressions.rs:94]: One way to improve these examples would be to bake in here the expected results

For example, using pretty_print on the collected result and asserting they are the same. 

Or maybe just copy/pasting the output as a comment

File: datafusion/physical-expr/src/regex_expressions.rs [hunk 10]
```diff
@@ -332,10 +482,70 @@ pub fn specialize_regexp_replace<T: OffsetSizeTrait>(
...
+        let patterns =
+            StringArray::from(vec!["^(a)", "^(A)", "(b|d)", "(B|D)", "^(b|c)"]);
+
+        let mut expected_builder: BooleanBuilder = BooleanBuilder::new();
+        expected_builder.append_value(true);
+        expected_builder.append_value(false);
+        expected_builder.append_value(true);
+        expected_builder.append_value(false);
+        expected_builder.append_value(false);
+        let expected = expected_builder.finish();
```
Comment (alamb) [inline datafusion/physical-expr/src/regex_expressions.rs:504]: This probably just follows the exisitng code in this file, but it is a bit more concise to build `BooleanArry`s like:

Something like this (untested):
```suggestion
        let expected = BooleanArray::from(vec![true, false, true, false, false]);
```

**Review by alamb — APPROVED** [review]
🤯  -- this PR is amazing. Thank you so much @Omega359 


The thoroughness of test coverage and documentation should serve as a model best practice. 

its-so-beautiful-crying-gif

File: docs/source/user-guide/sql/scalar_functions.md (hunk 4 above)
Comment (Omega359) [reply docs/source/user-guide/sql/scalar_functions.md]: I removed this from the PR

File: datafusion/physical-expr/src/regex_expressions.rs (hunk 9 above)
Comment (Omega359) [reply datafusion/physical-expr/src/regex_expressions.rs:94]: Good point, I can do that.

File: datafusion/physical-expr/src/regex_expressions.rs (hunk 10 above)
Comment (Omega359) [reply datafusion/physical-expr/src/regex_expressions.rs:504]: This is just me copy/pasting from some other bit of code and not thinking too much about it. I am still getting used to Rust's way of doing things :)

Comment (alamb) [reply datafusion/physical-expr/src/regex_expressions.rs:504]: I think following the existing structure is fine, I just figured I would point it out

File: datafusion/physical-expr/src/regex_expressions.rs (hunk 9 above)
Comment (Omega359) [reply datafusion/physical-expr/src/regex_expressions.rs:94]: completed.

File: datafusion-examples/examples/regexp.rs [hunk 11]
```diff
@@ -49,29 +43,83 @@ async fn main() -> Result<()> {
...
+        .await?
+        .collect()
         .await?;
 
-    // print the results
-    df.show().await?;
-
-    // use dataframe and regexp_like function to test col 'values', against patterns in col 'patterns' with flags
-    let df = ctx
+    assert_batches_eq!(
```
Comment (alamb) [inline datafusion-examples/examples/regexp.rs:56]: 👍

**Review by alamb — APPROVED** [review]

**Comment by alamb** [comment]
Thanks again @Omega359
